vending-app/src/main/kotlin/com/google/android/finsky/splitinstallservice/SplitInstallManager.kt

```diff
@@ -0,0 +1,337 @@
+        val sessionId: Int
+        var session: PackageInstaller.Session? = null
+        var totalDownloaded = 0L
+        try {
+            sessionId = packageInstaller.createSession(params)
+            session = packageInstaller.openSession(sessionId)
+            downloadList.forEach { item ->
+                val pkgPath = File(context.splitSaveFile().toString(), item.first)
+                session.openWrite(item.first, 0, -1).use { outputStream ->
+                    FileInputStream(pkgPath).use { inputStream -> inputStream.copyTo(outputStream) }
+                    session.fsync(outputStream)
+                }
+                totalDownloaded += pkgPath.length()
+                pkgPath.delete()
+            }
+            val deferred = CompletableDeferred<Intent>()
+            deferredMap[sessionId] = deferred
+            val intent = Intent(context, InstallResultReceiver::class.java).apply {
+                putExtra(KEY_BYTES_DOWNLOADED, totalDownloaded)
+            }
+            val pendingIntent = PendingIntent.getBroadcast(context, sessionId, intent, 0)
+            session.commit(pendingIntent.intentSender)
+            Log.d(TAG, "installSplitPackage session commit")
+            return deferred.await()
+        } catch (e: IOException) {
+            Log.w(TAG, "Error installing split", e)
+            throw e
+        } finally {
+            session?.close()
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private suspend fun downloadSplitPackage(context: Context, callingPackage: String, downloadList: ArraySet<Triple<String, String, Int>>): Boolean =
+        coroutineScope {
+            val results = downloadList.map { info ->
+                Log.d(TAG, "downloadSplitPackage: $info")
+                async {
+                    val downloaded = runCatching {
+                        httpClient.download(info.second, File(context.splitSaveFile().toString(), info.first), SPLIT_INSTALL_REQUEST_TAG)
+                    }.onFailure {
+                        Log.w(TAG, "downloadSplitPackage url:${info.second} save:${info.first}", it)
+                    }.getOrNull() != null
+                    downloaded.also { updateSplitInstallRecord(callingPackage, Triple(info.first, info.second, if (it) STATUS_DOWNLOADED else STATUS_UNKNOWN)) }
+                }
+            }.awaitAll()
+            return@coroutineScope results.all { it }
+        }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private suspend fun requestDownloadUrls(callingPackage: String, authToken: String, packs: MutableSet<String>): ArraySet<Triple<String, String, Int>> {
+        val versionCode = PackageInfoCompat.getLongVersionCode(context.packageManager.getPackageInfo(callingPackage, 0))
+        val requestUrl =
+            StringBuilder("https://play-fe.googleapis.com/fdfe/delivery?doc=$callingPackage&ot=1&vc=$versionCode&bvc=$versionCode&pf=1&pf=2&pf=3&pf=4&pf=5&pf=7&pf=8&pf=9&pf=10&da=4&bda=4&bf=4&fdcf=1&fdcf=2&ch=")
```

Comment:
I tried to execute the split installations from TikTok, but I wasn't able to download the split APKs until I changed the URL to:
```
https://play-fe.googleapis.com/fdfe/delivery?doc=$callingPackage&ot=1&vc=$versionCode
```
so removing the last part: 

```
&bvc=$versionCode&pf=1&pf=2&pf=3&pf=4&pf=5&pf=7&pf=8&pf=9&pf=10&da=4&bda=4&bf=4&fdcf=1&fdcf=2&ch=
```

What app did you use to test the split install feature ?

Reply:
When testing with TikTok, I face the following error;

```
SplitInstallManager: InstallResultReceiver onReceive: install fail -> INSTALL_FAILED_MISSING_SPLIT: Missing split for com.zhiliaoapp.musically
```

Reply:
@DaVinci9196 I was able to install a language split apk on PayPal :+1:

Reply:
> I tried to execute the split installations from TikTok, but I wasn't able to download the split APKs (the url triples were always null) until I changed the URL to:
> 
> ```
> https://play-fe.googleapis.com/fdfe/delivery?doc=$callingPackage&ot=1&vc=$versionCode
> ```
> 
> so removing the last part:
> 
> ```
> &bvc=$versionCode&pf=1&pf=2&pf=3&pf=4&pf=5&pf=7&pf=8&pf=9&pf=10&da=4&bda=4&bf=4&fdcf=1&fdcf=2&ch=
> ```
> 
> What app did you use to test the split install feature ?

Tested by switching languages ​​in Google Maps and Bolt